### PR TITLE
Disable the 'Publish' and 'Reject' buttons while the request is processing.

### PR DESCRIPTION
**Pull request by chuckharmston**
r? @ngokevin

### CODE REVIEW

**Comment by chuckharmston** [comment]
r-

We should just remove the item from state instead of setting it to say "published"

File: src/media/js/addon/components/addon.js [hunk 1]
```diff
@@ -55,6 +63,46 @@ export class Addon extends React.Component {
...
+
+  isButtonDisabled() {
+    return (this.props.isPublishing ||
+            this.props.isPublished ||
+            this.props.isRejected ||
+            this.props.isRejecting);
+  }
+
+  publishButton() {
+    let buttonText = () => {
```
Comment (ngokevin) [inline src/media/js/addon/components/addon.js]: why is this lazy? if we need a closure, we can do ```publishButton = () =>```

**Comment by chuckharmston** [comment]
re-r? @ngokevin 

Now it disables the reject/publish buttons when one is clicked, then reenables them if the request fails and deletes the queue item from the store when it's successfully rejected or published.

File: src/media/js/addon/actions/review.js [hunk 2]
```diff
@@ -38,10 +50,15 @@ export function publish(slug) {
       urlJoin(process.env.MKT_API_ROOT, `extensions/queue/${slug}/publish/`)
     ).q(apiArgs);
 
+    dispatch(publishPending(slug));
     req
       .post(publishUrl)
       .then((res, err) => {
-        dispatch(publishOk(slug));
+        if (res.status >= 400) {
```
Comment (ngokevin) [inline src/media/js/addon/actions/review.js]: why not just err

File: src/media/js/addon/components/addon.js [hunk 3]
```diff
@@ -55,6 +61,30 @@ export class Addon extends React.Component {
...
     return <h2>{this.props.name}</h2>
   }
+
+  isButtonDisabled() {
+    return (this.props.isPublishing ||
+            this.props.isRejecting);
+  }
+
+  publishButton() {
+    let buttonText = this.props.isPublishing ? 'Publishing...' : 'Publish';
```
Comment (ngokevin) [inline src/media/js/addon/components/addon.js]: you can use const

File: src/media/js/addon/components/addon.js [hunk 4]
```diff
@@ -55,6 +61,30 @@ export class Addon extends React.Component {
     }
     return <h2>{this.props.name}</h2>
   }
+
+  isButtonDisabled() {
+    return (this.props.isPublishing ||
+            this.props.isRejecting);
+  }
+
+  publishButton() {
```
Comment (ngokevin) [inline src/media/js/addon/components/addon.js]: i don't think this needs to be refactored into its own method yet, but if we are, i like to prefix methods that return components with ```render``` (e.g., renderPublishButton)

File: src/media/js/addon/actions/review.js (hunk 2 above)
Comment (chuckharmston) [inline src/media/js/addon/actions/review.js]: Copied what you did here: https://github.com/mozilla/marketplace-content-tools/blob/master/src/media/js/addon/actions/submit.js#L117

File: src/media/js/addon/containers/reviewDetail.js [hunk 5]
```diff
@@ -17,13 +17,13 @@ export class AddonReviewDetail extends React.Component {
   constructor(props) {
     super(props);
     // this.props.fetch();
-    this.props.addon = this.props.addons[this.props.slug];
   }
   render() {
+    let addon = this.props.addons[this.props.slug];
```
Comment (ngokevin) [inline src/media/js/addon/containers/reviewDetail.js]: const

File: src/media/js/addon/containers/reviewDetail.js [hunk 6]
```diff
@@ -17,13 +17,13 @@ export class AddonReviewDetail extends React.Component {
   constructor(props) {
     super(props);
     // this.props.fetch();
-    this.props.addon = this.props.addons[this.props.slug];
```
Comment (ngokevin) [inline src/media/js/addon/containers/reviewDetail.js]: thanks

File: src/media/js/addon/reducers/review.js [hunk 7]
```diff
@@ -17,6 +17,44 @@ export default function addonReviewReducer(state=initialState, action) {
...
 
+    case reviewActions.PUBLISH_PENDING: {
+      const newState = Object.assign({}, state);
+      newState[action.payload].isPublishing = true;
+      return newState;
+    }
+
+    case reviewActions.PUBLISH: {
+      const newState = Object.assign({}, state);
+      newState[action.payload].isPublishing = false;
```
Comment (ngokevin) [inline src/media/js/addon/reducers/review.js]: haha you're gonna delete it from the state anyways.

**Comment by ngokevin** [comment]
r+wc

**Pull request merged**